### PR TITLE
HPCC-25326 Remove some ambiguous xpath checking in release builds

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith For discussion/review.  A few questions:
* I debated using a different #define to allow the checking to be enabled in release mode if desired.  Would that be better?
* There are a couple of other similar checks, should they also be disabled?
* Which build to target?  The vast majority of the time saved on the original example is in my other fix, but this could affect xml reading/graph parsing/anything that accesses a IPropertyTree.

**Comment by richardkchapman** [comment]
Could argue this should go to 7.8?

**Comment by ghalliday** [comment]
@richardkchapman I think safe enough - so re-targeted.

**Comment by jakesmith** [comment]
The github smoketest failure is due to a bug in the smoketest.
The above result was triggered from when this PR targeted 7.12.x (really I think github should remove the non-7.12.x tests above from the list..)

But it failed because it triggered, and then the PR base changed, before the smoketest had started.
The action pull the branch by name, and got the 7.8.x version, which then failed because the smoketest isn't there (or in particular because one of the scripts it uses isn't there).

I'll open a separate JIRA to fix this smoketest issue.

**Comment by jakesmith** [comment]
@ghalliday is the indexcachebug mismatch genuine / related to this PR?

**Comment by jakesmith** [comment]
>I debated using a different #define to allow the checking to be enabled in release mode if desired. Would that be better?

I think it's sufficient to enable in DEBUG only.

> There are a couple of other similar checks, should they also be disabled?

yes, IMO, the other cases should also be disabled, they might not save as much, but better if it was consistent.

**Review by jakesmith — COMMENTED** [review]
@ghalliday - see comments, do you want to change other AMBIG. cases so that they are checked in DEBUG only too in this PR?

**Comment by ghalliday** [comment]
@jakesmith please check the new commit

**Review by jakesmith — APPROVED** [review]
@ghalliday - looks good

File: system/jlib/jptree.cpp [hunk 1]
```diff
@@ -1083,8 +1087,10 @@ void PTree::resolveParentChild(const char *xpath, IPropertyTree *&parent, IPrope
             }
             if (!parent)
             {
+#ifdef _DEBUG
```
Comment (jakesmith) [inline system/jlib/jptree.cpp:1090]: @ghalliday - as the compile failure (in Release) highlights, relating to 'multiplePaths' not used,
the check on 1084-1085 can also be avoided (wrapped in a _DEBUG), along with the declaration.

**Comment by ghalliday** [comment]
@jakesmith I think my change needs careful checking.

**Comment by ghalliday** [comment]
@richardkchapman I have double checked and I cannot see any potential changes in behaviour (other than losing ambiguous errors)

**Pull request merged**